usl_models/usl_models/flood_ml/model.py

```diff
@@ -376,75 +336,56 @@ def forward(
+        temporal = full_input["temporal"]
 
-        return predictions
+        B = spatiotemporal.shape[0]
+        C = 1
+        F, N, M = constants.GEO_FEATURES, constants.N_FLOOD_MAPS, constants.M_RAINFALL
+        T_MAX = constants.MAX_RAINFALL_DURATION
+        H, W = self._spatial_height, self._spatial_width
 
-    def _update_temporal_inputs(
-        self, flood_maps: tf.TensorArray, full_temp_input: tf.Tensor, t: int
-    ) -> tuple[tf.Tensor, tf.Tensor]:
-        """Updates temporal inputs for a new time step (prediction).
+        tf.ensure_shape(spatiotemporal, (B, N, H, W, C))
+        tf.ensure_shape(geospatial, (B, H, W, F))
+        tf.ensure_shape(temporal, (B, T_MAX, M))
 
-        Returns the appropriate rainfall and flood map inputs for the t-th
-        prediction. The number n of flood maps and rainfall windows returned is
-        the minimum of self._params.n_flood_maps and t.
-
-        Args:
-            flood_maps: A TensorArray of all flood maps.
-            full_time_input: The [T, m] tensor of all rainfall windows. This
-                function retrieves the appropriate windows given the time step.
-            t: Time step in the autoregression.
+        # This array stores the n predictions.
+        predictions = tf.TensorArray(tf.float32, size=n)
 
-        Returns:
-            A tuple of tensors for the flood maps and rainfall, having shapes
-            [n, H, W, 1] and [n, m], respectively.
-        """
-        n = min(self._params.n_flood_maps, t)
-        step_range = tf.range(t - n, t)
-        # Gather the relevant flood maps. This will stack them along axis 0, so
-        # we need to permute the time (0) and batch (1) axes.
-        st_input = flood_maps.gather(step_range)
-        st_input = tf.transpose(st_input, perm=[1, 0, 2, 3, 4])
-        # Gather the relevant rainfall windows.
-        temp_input = tf.gather(full_temp_input, step_range, axis=1)
-        return (st_input, temp_input)
+        # We use 1-indexing for simplicity. Time step t represents the t-th flood
+        # prediction.
+        for t in range(1, n + 1):
+            input = Input(geospatial=geospatial,
+                          temporal=self._get_temporal_window(temporal, t, N),
+                          spatiotemporal=spatiotemporal)
+            prediction = self.call(input)
+            predictions = predictions.write(t - 1, prediction)
+
+            # Append new predictions along time axis, drop the first.
+            spatiotemporal = tf.concat([spatiotemporal,
+                                        tf.expand_dims(prediction, axis=1)],
+                                       axis=1)[:, 1:]
+
+        # Gather dense tensor out of TensorArray along the time axis.
+        predictions = tf.stack(tf.unstack(predictions.stack()), axis=1)
+        # Drop channels dimension.
+        return tf.squeeze(predictions, axis=-1)
+
+    def _get_temporal_window(self, temporal: tf.Tensor, t: int, n: int) -> tf.Tensor:
```

Comment:
Is this a replacement for the `temporal_window_view` function in `data_utils.py`? If so, can you add more detail to the docstring explaining what this is doing? It would also be useful to add a test to make sure the behavior is WAI (and will help others understand what inputs/outputs might look like.

Reply:
Yes, I've added a unit test for this.

Reply:
Can you add details to the docstring explaining what the `temporal` tensor looks like when this is called?

Let's add the args and return section; I think it's better to have more detail than less here. The "n-sized window" is a bit confusing, because there are actually two temporal windows of interest in the model: the rainfall `m` and the spatiotemporal `n`. We're actually retrieving the `n` most recent (`m`-sized) rainfall windows at the current time step.

FWIW I wrote the original code and this still trips me up.

Reply:
Done.